src/interpolate.rs

```diff
@@ -0,0 +1,216 @@
+use {Duplex, Frame, Sample};
+
+/// An iterator that converts the rate at which frames are yielded from some given frame
+/// Interpolator into a new type.
+///
+/// Other names for `sample::rate::Converter` might include:
+///
+/// - Sample rate converter
+/// - {Up/Down}sampler
+/// - Sample interpolater
+/// - Sample decimator
+///
+#[derive(Clone)]
+pub struct Converter<T: Iterator, I: Interpolator>
+    where <T as Iterator>::Item: Frame
+{
+    source: T,
+    interpolator: I,
+    interpolation_value: f64,
+    source_to_target_ratio: f64
+}
+
+/// Interpolator that just rounds off any values to the previous value from the source
+pub struct Floor<F>
+{
+    left: F
+}
+
+/// Interpolator that interpolates linearly between the previous value and the next value
+pub struct Linear<F>
+{
+    left: F,
+    right: Option<F>
+}
+
+/// Trait for all things that can interpolate between two values. Implementations should keep track
+/// of the necessary data both before and after the current frame.
+pub trait Interpolator
+{
+    type Frame: Frame;
+
+    /// Given a distance between [0. and 1.) to the following sample, return the interpolated value
+    fn interpolate(&self, x: f64) -> Self::Frame;
+
+    /// Called whenever the Interpolator value is over 1.
+    fn next_source_frame(&mut self, source_frame: Option<Self::Frame>);
```

Comment:
Found it. It's because the Interpolator doesn't know about the stream, and (I believe) shouldn't return any value at all on a function that just swaps out internal variables.

I think this functionality (stopping an interpolator when it ends) should be in the Converter, which actually knows about the stream. Adding that now.